R/utilities-qualification-gof.R

```diff
@@ -0,0 +1,409 @@
+
+        simulationFile <- configurationPlan$getSimulationPath(
+          project = projectName,
+          simulation = simulationName
+        )
+
+        simulationResultsFile <- configurationPlan$getSimulationResultsPath(
+          project = projectName,
+          simulation = simulationName
+        )
+
+        simulation <- loadSimulation(simulationFile, loadFromCache = TRUE)
+        simulationResults <- importResultsFromCSV(simulation = simulation, filePaths = simulationResultsFile)
+
+        outputs <- lapply(simulation$outputSelections$allOutputs, function(output) {
+          Output$new(output$path)
+        })
+        names(outputs) <- lapply(simulation$outputSelections$allOutputs, function(output) {
+          output$path
+        })
+        output <- outputs[[outputPath]]
+        molWeight <- simulation$molWeightFor(outputPath)
+        simulationDimension <- getQuantity(path = outputPath, container = simulation)$dimension
+
+
+
+        # Setup simulations dataframe
+        simulatedDataStandardized <- data.frame(
+          Time = simulationResults$timeValues,
+          Concentration = simulationResults$getValuesByPath(path = outputPath, individualIds = 0)
+        )
+
+        # Setup observations dataframe
+        observedDataFileData <- read.csv(file.path(inputFolder, observedDataSetFilePath), check.names = FALSE, fileEncoding = "UTF-8-BOM")
+        observedDataFileMetaData <- parseObservationsDataFrame(observedDataFileData)
+        # If simulation in c("Amount","'"Concentration (molar)") and observations in c("Mass","'"Concentration (mass)"), convert observations to c("Amount","'"Concentration (molar)")
+        if (simulationDimension %in% c(ospDimensions$Amount, ospDimensions$`Concentration (molar)`)) {
+          observationsDimension <- massMoleConversion(ospsuite::getDimensionForUnit(observedDataFileMetaData$output$unit))
+        }
+        # Verify that simulations and observations have same dimensions
+        validateIsIncluded(values = observationsDimension, parentValues = simulationDimension, nullAllowed = FALSE)
+
+        observedDataStandardized <- observedDataFileData[, c(1, 2)]
+        names(observedDataStandardized) <- c("Time", "Concentration")
+        observedDataStandardized$Time <- ospsuite::toBaseUnit(
+          quantityOrDimension = ospDimensions$Time,
+          values = observedDataStandardized$Time,
+          unit = observedDataFileMetaData$time$unit
+        )
+        observedDataStandardized$Concentration <- ospsuite::toBaseUnit(
+          quantityOrDimension = observationsDimension,
+          values = observedDataStandardized$Concentration,
+          unit = observedDataFileMetaData$output$unit,
+          molWeight = molWeight
+        )
+
+
+        commonTimePoints <- intersect(observedDataStandardized$Time, simulatedDataStandardized$Time)
```

Comment:
Does `interest` capture all the data ? 
In residuals of RE, I am assuming that the time points may not perfectly match and that the simulation points are dense enough to be close to the observed data. Consequently, I am calculating the closest simulation time points to the observed time points in function `getResiduals` in utilities-goodness-of-fit.R
If it does I could replace the current method. Otherwise, you can leverage it if you want.

Reply:
I think the only robust way to deal with this is to read the time points in the observed data set and explicitly add them to the simulation before simulating.

eg
```
timepointsFromObsDataCSV <- c(0, 30, 60, 120, 240, 480, 960, 1440)
simulation$outputSchema$addTimePoints(timepointsFromObsDataCSV)
```

Then we can use `intersect` cleanly.

This will involve going through the dataset/simulation pairs and identifying the timepoints when setting up the workflow.